### PR TITLE
MOL2 parser populates elements attribute

### LINKED ISSUE #3062: MOL2 parser does not populate the elements attribute
## Is your feature request related to a problem? ##

The MOL2 parser does not populate the elements attribute.

## Describe the solution you'd like ##

I'd like the MOL2 parser to populate the elements attribute.

## Additional context ##

See #2553, #2421 and #2423 for context.

```
u = mda.Universe(mol2_molecule)
u.atoms.elements
```
throws a `MDAnalysis.exceptions.NoDataError`.

I'm working on a fix.

### PR DESCRIPTION
**Pull request by RMeli**
Fixes #3062 

Changes made in this Pull Request:
 - Add `elements` attribute when parsing MOL2 file

The elements are validated as for the `PDBParser` and empty records are used if the element is invalid.

When #2927 and #3057 are merged I can add a test for the selection and remove the bonds from the `mol2_wrong_elements` test variable.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: .github/workflows/gh-ci.yaml
Comment (IAlibay) [inline .github/workflows/gh-ci.yaml:1]: This might be part of our current offline discussion, but why are these showing up on the diff?

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (richardjgowers) [inline package/MDAnalysis/topology/MOL2Parser.py:170]: Putting the warning inside the loop will spam the user if they've got many weird types. Probably a better pattern is to accumulate weird types into a set and if any print after the loop.

**Review by richardjgowers — DISMISSED** [review]

File: .github/workflows/gh-ci.yaml
Comment (RMeli) [reply .github/workflows/gh-ci.yaml]: Because I'm still a git noob!

Should be fixed now.

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (RMeli) [reply package/MDAnalysis/topology/MOL2Parser.py:170]: Good point, I like this very much.

File: testsuite/MDAnalysisTests/topology/test_mol2.py
Comment (RMeli) [inline testsuite/MDAnalysisTests/topology/test_mol2.py:95]: Uncomment once #2927 is merged.

File: testsuite/MDAnalysisTests/topology/test_mol2.py
Comment (RMeli) [inline testsuite/MDAnalysisTests/topology/test_mol2.py:112]: Remove once #3057 is merged.

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (IAlibay) [inline package/MDAnalysis/topology/MOL2Parser.py:175]: I might be wrong here, but AFAIK warnings will only output one warning if encountering multiples of the same warning, i.e. your previous code and this one _should_ be equivalent in terms of number of outputs?

There's a performance cost between the two, but iirc I benchmarked this on a large PDB once and the cost wasn't noticeable (probably because there are much larger bottlenecks in the PDBReader).

Comment (IAlibay) [reply package/MDAnalysis/topology/MOL2Parser.py:175]: That being said, maybe it's cleaner to just have one output for all elements? (is this was @richardjgowers was saying above?))

Comment (RMeli) [reply package/MDAnalysis/topology/MOL2Parser.py:175]: Do you mean printing them out as a list?

Comment (IAlibay) [reply package/MDAnalysis/topology/MOL2Parser.py:175]: yeah, I'm not sure how it'll affect readability, but at least you've only got one warning instead of N bad elements warnings.

Comment (RMeli) [reply package/MDAnalysis/topology/MOL2Parser.py]:
```
Unknown elements found for some atoms: {'X', 'XX'}. These have been given an empty element record.
```

**Comment by RMeli** [comment]
I looked into the test failures and the ones for `test_bat` are due to the fact that atom types in the `Molecule_comments_header.mol2` file are not SYBYL atom types (the MOL2 file is generated by GAFF, according to @IAlibay , #2262). As suggested offline by @IAlibay , I think a good way forward is to avoid any guess for non-standard atom types (but keep parsing such files as we did before) and populate the element attribute only if SYBYL atom types are provided?

The test failure for `test_rdkit` is due to single test that checks `AttributeError` is thrown when the `elements` attribute is not present (which is no longer the case by default). Using `delattr` does not work (`AttributeError: can't delete attribute`).

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (IAlibay) [inline package/MDAnalysis/topology/MOL2Parser.py:167]: I _think_ for the sake of being safe, it might make sense to just read elements as is and then compare them against a list of SYBYL atom types. That way if someone uses a weird MOL2 scheme that isn't an SYBYL but has `.` in the atom types, then we can make sure we treat them as unknowns.

Comment (RMeli) [reply package/MDAnalysis/topology/MOL2Parser.py]: Good point. I added a `SYBYL2SYMB` dictionary for type to element conversion. I had to change some of the test `.mol2` files in order to conform to the standard (`S.O` and `S.O2` instead of `S.o` and `S.o2`). Should we support both cases?

**Review by IAlibay — CHANGES_REQUESTED** [review]
couple of quick comments

File: testsuite/MDAnalysisTests/data/mol2/Molecule.mol2
Comment (RMeli) [reply testsuite/MDAnalysisTests/data/mol2/Molecule.mol2:10]: This might be the only case for inconsistency, since all other strings after `.` are lowercase in the standard. My preference would be to support `S.o2` and `S.o` only as special case (probably coming from https://github.com/openbabel/openbabel/issues/599 ?) and eventually switch to `lower()` if other elements are affected by similar problems, but happy to be convinced otherwise.

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (RMeli) [inline package/MDAnalysis/topology/MOL2Parser.py:105]: What do you think about this? Given that the actual types are stored in the `types` attribute, I think the `elements` attribute should only contain actual elements (that can be readily converted to atomic numbers, for example). In such case, `Du.C` should be treated as carbon?

File: testsuite/MDAnalysisTests/data/mol2/Molecule.mol2
Comment (IAlibay) [reply testsuite/MDAnalysisTests/data/mol2/Molecule.mol2:10]: Yeah just supporting those special cases sounds good to me.

Comment (IAlibay) [reply testsuite/MDAnalysisTests/data/mol2/Molecule.mol2:10]: P.S. You'll probably want to revert the changes to these files, we might as well keep them as-is (as a record of how they were originally created). I'd also recommend a test that reads in a file with all the supported atom types + at least one unexpected one that way we can be sure that everything works as expected.

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (IAlibay) [reply package/MDAnalysis/topology/MOL2Parser.py:105]: Yeah I don't think I agree with placing `Du` in elements, I think our general approach is that dummy atoms are empty strings. If users want to use the types they can directly access the `atomtypes` attribute.

Regarding `Du.C`, I would consider that a dummy atom, therefore no element. Maybe @lilyminium has a different view here? I generally think that if folks have gone through the trouble of making something a dummy atom, then it should be treated as such.

"Any, Hal (halogen), Het (heteroatom) and Hev (heavy atoms)" - I'm not sure what you mean here, could you explain this?

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG:26]: I think this is an enhancement rather than a fix?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple more comments, mostly responses to questions.

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (IAlibay) [inline package/MDAnalysis/topology/MOL2Parser.py:162]: How big of a performance hit do we take here if all the elements are invalid? Would just an `if`/`else` be more effective? (I'm thinking if someone decides to read an entire gaff2 file then you'll end up hitting the `KeyError` every single time?)

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (IAlibay) [inline package/MDAnalysis/topology/MOL2Parser.py:103]: Can you add Elements to the attribute list in the docstring? Maybe with a comment about how it's read based on Sybyl atom types?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Nearly there I think, just a couple of extra things.

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (IAlibay) [reply package/MDAnalysis/topology/MOL2Parser.py:167]: Probably worth you testing this on your end too, but it looks like an if/else would be better here in cases where you're hitting a lot of unknown atom types.

Here's my benchmark:
```python
import string
import random

def try_catch(var, dicter):
    try:
        return dicter[var]
    except KeyError:
        return ''


def if_else(var, dicter):
    if var in dicter:
        return dicter[var]
    else:
        return ''


alphabet_numbers = {}
for i, letter in enumerate(string.ascii_letters):
    alphabet_numbers[letter] = i

random_ascii = [random.choice(string.ascii_letters) for x in range(100000)]
rand_digit = [random.choice(string.digits) for x in range(100000)]

# ascii pass only - all entries exist in input dictionary - both are ~ 21.5 ms
%timeit -r50 retval = [try_catch(var, alphabet_numbers) for var in rand_ascii]
%timeit -r50 retval = [if_else(var, alphabet_numbers) for var in rand_ascii]

# digits only - all entries do not exist in input dictionary
## ~ 55 ms
%timeit -r50 retval = [try_catch(var, alphabet_numbers) for var in rand_digit]
## ~ 21 ms
%timeit -r50 retval = [if_else(var, alphabet_numbers) for var in rand_digit]
```

Please do test it on your side too (my laptop is what you would call a potato 🙃 )

I'm really doing a "worst case scenario someone is passing a 10k atoms non-sybyl atomtypes file", which is kinda insane, but also I've seen crazier things be done....

File: package/MDAnalysis/topology/tables.py
Comment (IAlibay) [inline package/MDAnalysis/topology/tables.py:399]: That PDF doesn't seem to point to anything? (at least on my end)

File: testsuite/MDAnalysisTests/topology/test_mol2.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/topology/test_mol2.py:35]: should these be moved about the MDA imports (I feel like theres a PEP for this but I can never remember), also shouldn't there be two lines between imports and the class (might be forgetting my PEP8)?

File: testsuite/MDAnalysisTests/topology/test_mol2.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/topology/test_mol2.py:111]: Might have missed it, but could you add a test that satisfies line 189 on the parser? i.e. all elements have not been assigned so the elements attribute is not created?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple more things, we should be good to merge after this.

File: package/MDAnalysis/topology/tables.py
Comment (RMeli) [reply package/MDAnalysis/topology/tables.py:399]: Odd, I can't access it anymore... Does the following work? [http://chemyang.ccnu.edu.cn/ccb/server/AIMMS/mol2.pdf](http://chemyang.ccnu.edu.cn/ccb/server/AIMMS/mol2.pdf)

Comment (IAlibay) [reply package/MDAnalysis/topology/tables.py:399]: Yeah that works for now. I think that technically the sybyl manual is copyrighted, not sure how to go about this 🤔 @orbeckst any ideas?

Comment (orbeckst) [reply package/MDAnalysis/topology/tables.py:399]: See if it's archived in the internet archive and link to it there. 

WebCITE does not accept new submissions anymore, I think.

Ideally, *someone* puts it on zenodo or figshare so that we have a DOI. Perhaps ask MolSSI to archive it as an important document.

Comment (IAlibay) [reply package/MDAnalysis/topology/tables.py:399]: @RMeli For the sake of getting this merged quickly, how about we raise an issue about adding details about the mol2 format manual, remove the link for now and then we can go bug the right person so we can link to it in the most appropriate way?

Comment (orbeckst) [reply package/MDAnalysis/topology/tables.py:399]: I don't think the link is inappropriate at all, it resolves — and that's all a user cares about. As long as someone else is hosting it...

Long-term a better/archive-grade link would be good. Raise an issue for that but keep the link!

Comment (orbeckst) [reply package/MDAnalysis/topology/tables.py:399]: I mean, put in the working link from https://github.com/MDAnalysis/mdanalysis/pull/3063#discussion_r687938122

Comment (IAlibay) [reply package/MDAnalysis/topology/tables.py:399]: 👍🏽 sounds good!

File: testsuite/MDAnalysisTests/topology/test_mol2.py
Comment (RMeli) [reply testsuite/MDAnalysisTests/topology/test_mol2.py:111]: Well spotted, thanks! There was indeed a bug due to the fact that `validated_elements` is a list and not a `np.ndarray`... I fixed it by pre-allocating `validated_elements` as an array (and added the corresponding test).

File: package/MDAnalysis/topology/MOL2Parser.py
Comment (RMeli) [reply package/MDAnalysis/topology/MOL2Parser.py:169]: Resolving conversation as duplicate.

**Comment by RMeli** [comment]
@IAlibay I could see your if/else vs try/catch benchmark twice but now that I resolved one conversation both disappeared. Anyways, I can reproduce your numbers locally: similar performance when no exception is triggered but a factor of 2 in performance drop when exceptions are triggered often. I'll change the code as suggested, thanks for bringing this up!

**Comment by IAlibay** [comment]
I'll have a final look in the morning - hopefully we should be good.

@richardjgowers - do you want to have another look or shall I dismiss your review? (I think your initial comments have all be addressed now)

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG:120]: Is it SYMBL or SYBYL atom types? (just looking at the manual, I thought it was the latter - pages 52-53?)

```suggestion
  * MOL2 parser populates elements attribute from SYBYL atom types (Issue #3062)
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just the one final thing 🙃

**Review by IAlibay — APPROVED** [review]

**Pull request merged**
